// pxkit/src/react/components/ImgButton.jsx
import React from "react";
import {handleImgBtnOut, handleImgBtnOver, getImgBtnPath} from "../../core/utils/dom";
import PropTypes from "prop-types";


/* dom Type
 * @Description:  普通的图片按钮 （normal，over，down状态下切换图片)，就一个图片元素
 * @Author: 彭祥 (Email:245803627@qq.com)
 * @Date: 2020/3/20 16:49
 * @LastEditors: pengxiang
 * @LastEditTime: 2020/3/20 16:49
 */
export default class ImgButton extends React.Component{
    static propTypes = {
        src: PropTypes.string,  //图片正常状态
        onDown: PropTypes.func, //button点击
        onUp: PropTypes.func,
        actionEnabled: PropTypes.bool, //此按钮是否被禁用交互
    };
    static defaultProps = {
        src: '',
        onDown: ()=>{},
        onUp: ()=>{},
        actionEnabled: false,
    };

    constructor(props) {
        super(props);
        this.state = {btnState: 'out', actionEnabled:props.actionEnabled};
        // this.actionEnabled = false;
    }

    componentDidMount() {
    }
    componentWillUnmount() {
    }


    handleImgBtnOver(event) {
        handleImgBtnOver(event, 'up', 'down');
        this.setState({btnState: 'over'});
    }
    handleImgBtnOut(event) {
        handleImgBtnOut(event, 'down', 'up');
        this.setState({btnState: 'out'});
    }
    handleImgDown(event) {
        if(this.state.actionEnabled === true){return false;}
        this.props.onDown(event.currentTarget, this);
    }
    handleImgUp(event) {
        if(this.state.actionEnabled === true){return false;}
        this.props.onUp(event.currentTarget, this);
    }

    setEnabled(bool) {
        // this.actionEnabled = bool;
        this.setState({actionEnabled: bool});
    }

    render() {
        return(
            <img className={this.props.className} id={this.props.id} style={this.props.style}
                 src={this.state.btnState === 'over' ? getImgBtnPath(this.props.src, 'up', 'down') : getImgBtnPath(this.props.src, 'down', 'up')}
                 title={this.props.title} alt={this.props.alt || ''}
                 onMouseOver={this.handleImgBtnOver.bind(this)}
                 onMouseOut={this.handleImgBtnOut.bind(this)}
                 onMouseDown={this.handleImgDown.bind(this)}
                 onMouseUp={this.handleImgUp.bind(this)}>
            </img>
        )
    }
}
