// pxkit/src/core/utils/index.js
/*
 * @Description:
 * @Author: 彭祥 (Email:245803627@qq.com)
 * @Date: 2020-09-02 16:28
 * @LastEditors: 24580
 * @LastEditTime: 2020-09-02 16:28
 */
// export * from './core';

/**
 * 屏蔽 游览器 右键菜单
 */
export const forbidBrowserMenu = () => {
    document.oncontextmenu = function() {return false;};
}














/*
 * 获取 年月日 时间
 * @returns {*[]} 一个年月日的数组[年，月，日]
 */
export function getYMDTime() {
    let nTime = [];
    const nowData = new Date();
    const year = getPrefixNum(nowData.getFullYear());
    const month = getPrefixNum(nowData.getMonth() + 1);
    const day = getPrefixNum(nowData.getDate());
    // nTime = "[" + year + ":" + month + ":" + day + "]   ";
    nTime = [year, month, day];
    return nTime;
}

/*获取时间 格式为 时分秒 [17:17:06]*/
export function getHMSTime() {
    var str = "";
    var mydata = new Date();
    var hour = getPrefixNum(mydata.getHours());
    var minute = getPrefixNum(mydata.getMinutes());
    var second = getPrefixNum(mydata.getSeconds());
    str = "[" + hour + ":" + minute + ":" + second + "]   ";
    return str;
}
function getPrefixNum(param1) {
    if (param1 < 10 && param1 >= 0) {
        return "0" + param1.toString();
    }
    return param1.toString();
}


/*
 * 把字符串转成对象Object
 * @param str [str='{name:"Helen",age:"22",sex:"female"}' 或者 str='id:zqcm, startNum:0, endNum:25']
 */
export function convertStr2Obj(str) {
    // const obj = new Function('return '+str)();
    str = str.replace('{', '').replace('}', '').replace(/\s*/g,"");
    const obj = new Object();
    const strs = str.split(",");
    for(let i = 0; i < strs.length; i ++) {
        obj[strs[i].split(":")[0]]=(strs[i].split(":")[1]);
    }
    return obj;
}


/*判断一个字符串是否是JSON格式（http://www.cnblogs.com/lanleiming/p/7096973.html）*/
export function isJSON(str) {
    if (typeof str == 'string') {
        try {
            var obj=JSON.parse(str);
            if(typeof obj == 'object' && obj ){
                return true;
            }else{
                return false;
            }

        } catch(e) {
            console.log('It is not a JSON!' +'\n'+'error：'+str +'\n'+e);
            return false;
        }
    }
    // console.log('It is not a string!');
}
